scripts/mkimg-boot-zynq7000.sh

```diff
@@ -0,0 +1,172 @@
+	$((OFFS_IMGH / 4))  # Address to the image header
+	0x00000000
+)
+
+IMG_HEADER=(
+	0x0                 # Next image header, 0 if last
+	$((OFFS_PARTH / 4)) # First partition header offset (words)
+	0x00000000          # Reserved
+	0x00000001          # Partition count length
+	0x00000000          # Image name
+	0x00000000          # String terminator
+)
+
+PARTITION_HEADER=(
+	$((SIZE_IMG / 4))  # Encrypted partition length (words)
+	$((SIZE_IMG / 4))  # Unencrypted partition length (words)
+	$((SIZE_IMG / 4))  # Total partition length (words)
+	0x00000000         # Destination load address to which to copy image
+	"$ADDR_ENTRY"      # Entry point of the partition
+	$((OFFS_IMG / 4))  # Data offset in the image (words)
+	0x00000013         # Attribute bits
+	0x00000001         # Section count
+	0x00000000         # Location of the checksum word
+	$((OFFS_IMGH / 4)) # Image header offset (words)
+	0x00000000         # Authentication certification word offset
+	0x00000000         # Reserved
+	0x00000000         # Reserved
+	0x00000000         # Reserved
+	0x00000000         # Reserved
+)
+
+
+#
+# Auxiliary functions
+#
+
+
+reverse() {
+	local array=("$@")
+	for data in "${array[@]}"; do
+		for (( i=0; i<4; ++i )); do
+			printf "%02x" $((data & 0xff))
+			data=$((data >> 8))
+		done
+	done
+}
+
+
+reverse_rpt() {
+	for (( j=0; j<$2; ++j )); do
+		reverse "$1"
+	done
+}
+
+
+checksum() {
+	local checksum=0
+	local array=("$@")
+
+	for data in "${array[@]}"; do
+		checksum=$((checksum + data))
+	done
+
+	checksum=$((checksum ^ 0xffffffff))
+	reverse $checksum
+}
+
+
+#
+# Make Boot Image Layout and add it at the beginning of the image
+#
+
+make_boot_layout() {
+	local filename="boot_layout"
+
+	# Save layout to the hex file. Empty spaces between headers should be filled with 0xff.
+	{
+		# Boot Header
+		reverse_rpt 0xeafffffe 8    # Dummy ARM Vector Table
+		reverse "${BOOT_HEADER[@]}"
+		checksum "${BOOT_HEADER[@]}"
+		reverse_rpt 0x00000000 19   # User defined fields
+		reverse "$OFFS_DEVH"
+		reverse "$OFFS_PARTH"
+
+		# Register Initialization Table
+		for (( j=0; j<256; ++j)); do
+			reverse 0xffffffff
+			reverse 0x00000000
+		done
+		reverse_rpt 0xffffffff $(((OFFS_DEVH - 256 * 4 * 2 - 0xa0) / 4 ))
+
+		# Device Image Header Table
+		reverse "${DEV_TABLE[@]}"
+		reverse_rpt 0xffffffff $(((OFFS_IMGH - (OFFS_DEVH + ${#DEV_TABLE[@]} * 4)) / 4))
+
+		# Image Header
+		reverse "${IMG_HEADER[@]}"
+		reverse_rpt 0xffffffff $(((OFFS_PARTH - (OFFS_IMGH + ${#IMG_HEADER[@]} * 4)) / 4))
+
+		# Partition Header
+		reverse "${PARTITION_HEADER[@]}"
+		checksum "${PARTITION_HEADER[@]}"
+		reverse_rpt 0xffffffff $(((OFFS_IMG - (OFFS_PARTH + (${#PARTITION_HEADER[@]} + 1) * 4)) / 4))
+
+	} > "${filename}.hex"
+
+	xxd -r -p "${filename}.hex" > "${filename}.bin"
+
+	# Add boot image layout to the corresponding image
+	cat "$PATH_IMG" >> "${filename}.bin"
+	mv "${filename}.bin" "$PATH_IMG"
```

Comment:
Output file should be different then the input - it will cause problems if the script is run again without clearing previous result.

Reply:
With `mv -f` to force overwrite,  the last `rm -f` is redundant.

Reply:
What I mean - we might have already have prepared PATH_IMG file from previous run - if we run script on it again, it will append boot header to file that already have this header - image will be invalid.

Reply:
Right.